runtime/etc/mli/mli_common_code.c

```diff
@@ -26,25 +26,91 @@
 #include <string.h>
 #include <stdarg.h>
 #include <assert.h>
+#include <unistd.h>
 #include <zmq.h>
 
+//
+// Grab headers for shims used by the client and server.
+//
+// The Chapel runtime is sanitized during Travis testing, and any calls to
+// LIBC functions are reported as errors. Sadly, our code gets sanitized too,
+// even though it _technically_ isn't part of the runtime itself. That's the
+// reason why any of these shims exist in the first place.
+//
+// We really don't want to call "chpl_exit_any()" in the client - we'd prefer
+// the LIBC "exit()" function. Fortunately, the LAUNCHER already accounts for
+// this. So we "gain access" to "exit()" by pretending to be the launcher.
+// This defines "chpl_exit_any()" as a macro representing "exit()".
+//
+#ifdef CHPL_MLI_IS_SERVER
+# include "chpl-mem.h"
+# include "error.h"
+#elif defined(CHPL_MLI_IS_CLIENT)
+# ifndef LAUNCHER
+#   define CHPL_MLI_LAUNCHER_IS_DEFINED
+#   define LAUNCHER
+# endif
+# include "chpl-mem-sys.h"
+# include "chplexit.h"
+# ifdef CHPL_MLI_LAUNCHER_IS_DEFINED
+#   undef CHPL_MLI_LAUNCHER_IS_DEFINED
+#   undef LAUNCHER
+# endif
+#endif
+
 //
 // If this code is being run on the server, mli_malloc() is a wrapper for
 // chpl_malloc(). If this code is being run on the client, then it is a
 // wrapper for the system allocator.
 //
 #ifdef CHPL_MLI_IS_SERVER
-# include "chpl-mem.h"
 # define mli_malloc(bytes) chpl_malloc(bytes)
 # define mli_free(ptr) chpl_free(ptr)
 #elif defined(CHPL_MLI_IS_CLIENT)
-# include "chpl-mem-sys.h"
 # define mli_malloc(bytes) sys_malloc(bytes)
 # define mli_free(ptr) sys_free(ptr)
 #else
 # error The mli_malloc/free macros were defined outside of client/server.
 #endif
 
+//
+// Terminate all the locales properly if the server, else just "exit()".
+//
+#ifdef CHPL_MLI_IS_SERVER
+# define mli_terminate() chpl_error("", 0, 0)
+#elif defined(CHPL_MLI_IS_CLIENT)
+# define mli_terminate() chpl_exit_any(1)
+#else
+# error The mli_terminate macro was defined outside of client/server.
+#endif
+
+//
+// Print prefix, used to indicate whether client or server is printing.
+//
+#ifdef CHPL_MLI_IS_SERVER
+# define chpl_mli_pfx "[Server]"
+#elif defined(CHPL_MLI_IS_CLIENT)
+# define chpl_mli_pfx "[Client]"
+#else
+# error The chpl_mli_pfx macro was defined outside of client/server.
+#endif
+
+//
+// Print a debug message, or do nothing if debug messages are turned off.
+//
+#ifdef CHPL_MLI_DEBUG_PRINT
+# ifndef chpl_mli_pfx
+#   error The chpl_mli_pfx macro must be defined.
+# endif
+# define chpl_mli_debugf(fmt, ...)                \
```

Comment:
I might define this in a header file that can get used in multiple places instead of shifting it and the .c #includes around multiple times.

Reply:
I think it's probably ok for now? If either of us has to move it around anymore then I can definitely move it to a header file.